Gems/Atom/RHI/Code/Include/Atom/RHI/Allocator.h

```diff
@@ -86,44 +83,42 @@ namespace AZ
-             */
+            //! Allocates a virtual address relative to the base address provided at initialization time.
+            //! @param byteCount The number of bytes to allocate.
+            //! @param byteAlignement The alignment used to align the allocation.
             virtual VirtualAddress Allocate(size_t byteCount, size_t byteAlignment) = 0;
 
-            /**
-             * Deallocates an allocation. The memory is not reclaimed until garbage collect is called.
-             * Depending on the garbage collection latency, it may take several garbage collection cycles
-             * before the memory is reclaimed.
-             */
+            //! Deallocates an allocation. The memory is not reclaimed until garbage collect is called.
+            //! Depending on the garbage collection latency, it may take several garbage collection cycles
+            //! before the memory is reclaimed.
             virtual void DeAllocate(VirtualAddress offset) = 0;
 
-            /// Allocations are deferred-released until a specific number of GC cycles have occurred. This
-            /// is useful for allocations actively being consumed by the GPU.
+            //! Allocations are deferred-released until a specific number of GC cycles have occurred. This
+            //! is useful for allocations actively being consumed by the GPU.
             virtual void GarbageCollect() = 0;
 
-            /// Forces garbage collection of all allocations, regardless of the GC latency.
+            //! Forces garbage collection of all allocations, regardless of the GC latency.
             virtual void GarbageCollectForce() = 0;
 
-            /**
-             * Returns the number of allocations active for this allocator. This includes
-             * allocations that are pending garbage collection.
-             */
+            //! Returns the number of allocations active for this allocator. This includes
+            //! allocations that are pending garbage collection.
             virtual size_t GetAllocationCount() const { return 0; }
 
-            /**
-             * Returns the number of bytes used by the allocator. This includes
-             * allocations that are pending garbage collection.
-             */
+            //! Returns the number of bytes used by the allocator. This includes
+            //! allocations that are pending garbage collection.
             virtual size_t GetAllocatedByteCount() const { return 0; }
 
-            /// Returns the descriptor used to initialize the allocator.
+            //! Returns the descriptor used to initialize the allocator.
             virtual const Descriptor& GetDescriptor() const = 0;
 
-            /// Helper for converting agnostic VirtualAddress type to pointer type. Will convert
-            /// VirtualAddress::Null to nullptr.
+            //! Clone the current allocator to the new allocator passed in
+            virtual void Clone([[maybe_unused]] RHI::Allocator* newAllocator)
```

Comment:
I think we should consider using [CRTP](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern#:~:text=The%20curiously%20recurring%20template%20pattern,form%20of%20F%2Dbounded%20quantification.) here to make this method actually typesafe. For example:

```cpp
template <typename T>
class Allocator
{
    // ...
    void Clone(T& newAllocator) { /*...*/ }
};

class FreeListAllocator : public Allocator<FreeListAllocator>
{
    // ...
    void Clone(FreeListAllocator& newAllocator) override { /*...*/ }
};
```

Personally, I prefer using a reference type there also since the argument isn't optional/nullable

Reply:
I like the idea behind this pattern but the problem with it is that DescriptorPool doesnt really know what allocator it will use at compile time. It is only at runtime that it decides what allocator type to use. I am not sure how to easily convert that using the CRTP pattern. For example based on certain settings it could invoke either of the code below

RHI::FreeListAllocator* allocator = aznew RHI::FreeListAllocator();
allocator->Init(descriptor);
m_allocator.reset(allocator);

                or 

RHI::PoolAllocator* allocator = aznew RHI::PoolAllocator();
allocator->Init(descriptor);
m_allocator.reset(allocator);